src/include/frag.h

```diff
@@ -211,6 +211,8 @@ class fragtree_t {
 	ls.front() == x &&
 	ls.size() == 1)
       return true;
+    // A node without leaves underneath it is definitely a leaf... 
+    if (ls.empty()) return true;
```

Comment:
I ran into a couple of issues because I was doing weird migrations, like splitting a single directory and moving individual dirfrags multiple times (e.g., from MDS0 to MDS1 to MDS4). 

In this case, when individual dirfrags landed on new MDSs, the client was prematurely deleting them from its fragmap ([here](https://github.com/ceph/ceph/blob/master/src/client/Client.cc#L730)). Before this change, it would delete the dirfrag because the call to is_leaf() returned false. I figured a single dirfrag without anything underneath it was a leaf anyways?

Another example of overlooking migrations with individual fragments is in [try_rebalance()](https://github.com/ceph/ceph/blob/master/src/mds/MDBalancer.cc#L598). When trying to decide which subtrees to move, the balancer traverses over its subtrees starting at the root directories... but if it's just managing a bunch of individual dirfrags, it ignores them. That's why I added [this](https://github.com/ceph/ceph/pull/5155/files#diff-de652aa7a28959252eb4486fdcbab7feR1069) piece of code.